cmp/clock/Clock.js

```diff
@@ -0,0 +1,135 @@
+import {isEmpty, isNil, isNumber} from 'lodash';
+import moment from 'moment';
+import PT from 'prop-types';
+
+/**
+ * A component to display the current time.
+ *
+ * Automatically updates on a regular interval to stay current.
+ */
+export const [Clock, clock] = hoistCmp.withFactory({
+    displayName: 'Clock',
+    className: 'xh-clock',
+
+    render({timezone, format, updateInterval, prefix, suffix, ...props}) {
+        const impl = useLocalModel(LocalModel);
+        impl.setData({timezone, format, updateInterval, prefix, suffix});
+
+        return box({
+            ...props,
+            item: span(impl.display)
+        });
+    }
+});
+Clock.propTypes = {
+    /**
+     * Timezone id. The user's local system time will be used if omitted.
+     * @see {@link https://docs.oracle.com/middleware/12212/wcs/tag-ref/MISC/TimeZones.html} */
+    timezone: PT.string,
+
+    /** moment.js date format string. @see {@link https://momentjs.com/docs/#/displaying/format/} */
+    format: PT.string,
+
+    /** Prefix for the displayed text */
+    prefix: PT.string,
+
+    /** Suffix for the displayed text */
+    suffix: PT.string,
+
+    /** Frequency (ms) for updating the displayed text */
+    updateInterval: PT.number
+};
+
+@HoistModel
+class LocalModel {
+    timezone;
+    format;
+    updateInterval;
+    prefix;
+    suffix;
+
+    offset;
+    offsetException;
+
+    @observable display = '';
+
+    @managed
+    timer = Timer.create({
+        runFn: () => this.refreshDisplay(),
+        interval: ONE_SECOND
+    });
+
+    setData({timezone, format = 'h:mm:ss a', updateInterval = ONE_SECOND, prefix, suffix}) {
+        if (timezone !== this.timezone) {
+            this.timezone = timezone;
+            this.updateTimezoneAsync();
+        }
+
+        this.format = format;
+        this.updateInterval = updateInterval;
+        this.prefix = prefix;
+        this.suffix = suffix;
+
+        this.timer.setInterval(updateInterval);
+    }
+
+    async updateTimezoneAsync() {
+        const {timezone} = this;
+        if (!isEmpty(timezone)) {
+            try {
+                const {offset} = await XH.fetchJson({url: 'xh/getTimeZoneOffset', params: {timeZoneId: timezone}});
+                this.offset = offset;
+            } catch (e) {
+                XH.handleException(e, {showAlert: false});
+                this.offset = null;
+                this.offsetException = e;
+            } finally {
+                this.refreshDisplay();
+            }
+        } else {
+            this.offset = null;
+            this.offsetException = null;
+            this.refreshDisplay();
+        }
+    }
+
+    @action
+    refreshDisplay() {
+        if (this.offsetException) {
+            this.display = '';
+            return;
+        }
+
+        const {prefix, suffix, format, offset} = this,
+            time = getAdjustedTime(offset),
+            parts = [time ? moment(time).format(format) : '???'];
```

Comment:
Looks like we will always have a time here so don't think this will fall through.

Should we reserve the `???` display for the (hopefully rare!) case where the client is unable to load the offset and we have an exception, above?  Maybe still pushing on the prefix / suffix?